File: examples/python/raw_bidi_client.py
import asyncio
import json
import os

import websockets


async def main() -> None:
    endpoint = os.environ.get("CAMOUFOX_BIDI_ENDPOINT")
    if not endpoint:
        raise SystemExit("set CAMOUFOX_BIDI_ENDPOINT, for example ws://127.0.0.1:50123/session")

    async with websockets.connect(endpoint) as ws:
        await send(ws, 1, "session.status", {})
        print("status:", await read_result(ws, 1))

        await send(ws, 2, "session.new", {"capabilities": {}})
        await read_result(ws, 2)

        await send(ws, 3, "browsingContext.create", {"type": "tab"})
        created = await read_result(ws, 3)
        context = created["context"]

        await send(
            ws,
            4,
            "browsingContext.navigate",
            {
                "context": context,
                "url": "data:text/html,<title>go-camoufox</title><h1>hello from Python</h1>",
                "wait": "complete",
            },
        )
        await read_result(ws, 4)

        await send(
            ws,
            5,
            "script.evaluate",
            {
                "expression": "document.querySelector('h1').textContent",
                "target": {"context": context},
                "awaitPromise": True,
            },
        )
        evaluated = await read_result(ws, 5)
        print(evaluated["result"]["value"])

        await send(ws, 6, "browsingContext.close", {"context": context})
        await read_result(ws, 6)
        await send(ws, 7, "session.end", {})
        await read_result(ws, 7)


async def send(ws, command_id: int, method: str, params: dict) -> None:
    await ws.send(json.dumps({"id": command_id, "method": method, "params": params}))


async def read_result(ws, command_id: int) -> dict:
    while True:
        message = json.loads(await ws.recv())
        if message.get("id") != command_id:
            continue
        if message.get("type") != "success":
            raise RuntimeError(json.dumps(message, indent=2))
        return message.get("result", {})


if __name__ == "__main__":
    asyncio.run(main())

File: examples/node/raw_bidi_client.mjs
import WebSocket from "ws";

const endpoint = process.env.CAMOUFOX_BIDI_ENDPOINT;
if (!endpoint) {
  throw new Error("set CAMOUFOX_BIDI_ENDPOINT, for example ws://127.0.0.1:50123/session");
}

const ws = new WebSocket(endpoint);
const pending = new Map();

ws.on("message", (data) => {
  const message = JSON.parse(data.toString());
  if (!message.id || !pending.has(message.id)) return;
  const { resolve, reject } = pending.get(message.id);
  pending.delete(message.id);
  if (message.type === "success") resolve(message.result ?? {});
  else reject(new Error(JSON.stringify(message, null, 2)));
});

await once(ws, "open");

let nextId = 1;
const command = (method, params = {}) => {
  const id = nextId++;
  ws.send(JSON.stringify({ id, method, params }));
  return new Promise((resolve, reject) => pending.set(id, { resolve, reject }));
};

console.log("status:", await command("session.status"));
await command("session.new", { capabilities: {} });

const created = await command("browsingContext.create", { type: "tab" });
const context = created.context;

await command("browsingContext.navigate", {
  context,
  url: "data:text/html,<title>go-camoufox</title><h1>hello from Node</h1>",
  wait: "complete",
});

const evaluated = await command("script.evaluate", {
  expression: "document.querySelector('h1').textContent",
  target: { context },
  awaitPromise: true,
});

console.log(evaluated.result.value);
await command("browsingContext.close", { context });
await command("session.end");
ws.close();

function once(emitter, event) {
  return new Promise((resolve, reject) => {
    emitter.once(event, resolve);
    emitter.once("error", reject);
  });
}
